### PR TITLE
chore: debug 0.0.11 synchronization

### PR DESCRIPTION
**Pull request by Freyskeyd**
# Description

This PR explores the issue with the synchronization of a new node. It solves the issue by refactoring a bit the calculation of diff between local and input.
The problem relied on the `prefix_iter_at` method that was introduced during the synchronization phase, but didn't work as expected in some cases. A ticket has been created to fix or refactor it, in the meantime, the synchronization process is falling back to the `skip` method.

## PR Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added or updated tests that comprehensively prove my change is effective or that my feature works

### CODE REVIEW

File: crates/topos-tce-storage/src/tests/checkpoints.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    for cert in certificates_a {
+        _ = store.insert_certificate_delivered(&cert).await;
+    }
+
+    for cert in certificates_b {
+        _ = store.insert_certificate_delivered(&cert).await;
+    }
+
+    let checkpoint = store
+        .get_checkpoint_diff(vec![checkpoint])
```
Comment (atanmarko) [inline crates/topos-tce-storage/src/tests/checkpoints.rs:100]: Maybe `get_checkpoint_diff` could take references as a function - no need then to clone whole vectors?

**Review by atanmarko — APPROVED** [review]

File: crates/topos-tce-synchronizer/src/checkpoints_collector/mod.rs [hunk 2]
```diff
@@ -184,11 +189,11 @@ impl CheckpointSynchronizer {
         diff: HashMap<SubnetId, Vec<ProofOfDelivery>>,
     ) -> Result<Vec<Vec<CertificateId>>, SyncError> {
         let mut certs: HashSet<CertificateId> = HashSet::new();
-        for (_subnet, proofs) in diff {
+        for (subnet, proofs) in diff {
             let len = proofs.len();
             let unverified_certs = self.store.insert_unverified_proofs(proofs)?;
 
-            debug!("Persist {} unverified proofs", len);
+            debug!("Persist {} unverified proofs for {}", len, subnet);
```
Comment (hadjiszs) [inline crates/topos-tce-synchronizer/src/checkpoints_collector/mod.rs:196]:
```suggestion
            debug!("Persist {} unverified proof of delivery for {}", len, subnet);
```

File: crates/topos-tce-broadcast/src/double_echo/broadcast_state.rs [hunk 3]
```diff
@@ -230,13 +230,20 @@ impl BroadcastState {
...
         {
             Some(consumed) => consumed >= self.delivery_threshold,
             None => false,
-        }
+        };
+
+        debug!(
+            "📝 Certificate {} reached Delivery threshold: {}",
+            &self.certificate.id, delivery_threshold
+        );
```
Comment (hadjiszs) [inline crates/topos-tce-broadcast/src/double_echo/broadcast_state.rs:245]: This one will print out often and redundant with logs in `update_status` no? Maybe to remove or to put it as `trace`?

**Review by hadjiszs — APPROVED** [review]

File: crates/topos-tce-broadcast/src/double_echo/broadcast_state.rs (hunk 3 above)
Comment (Freyskeyd) [reply crates/topos-tce-broadcast/src/double_echo/broadcast_state.rs:245]: It matches what we have for `echo` and `ready` threshold, when the certificate is delivered, even if we have more message coming, they're ignored

File: crates/topos-tce-storage/src/tests/checkpoints.rs (hunk 1 above)
Comment (Freyskeyd) [reply crates/topos-tce-storage/src/tests/checkpoints.rs]: I updated the `get_checkpoint_diff` to take a ref

File: crates/topos-tce-storage/src/tests/checkpoints.rs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    );
+    assert_eq!(
+        *checkpoint.get(&SOURCE_SUBNET_ID_2).unwrap(),
+        (0..=23).collect::<Vec<_>>()
+    );
+}
+
+#[rstest]
+#[tokio::test]
+async fn get_checkpoint_diff_with_input(store: Arc<ValidatorStore>) {
```
Comment (dvdplm) [inline crates/topos-tce-storage/src/tests/checkpoints.rs:84]: I do not understand what this test actually does, can you explain please? The part I don't get is what the "input" part is. The only difference between this test and the "no_input" version seems to be that this one does `.get(20)` and clones the proof of delivery, but I don't see how that maps to "with input".

File: crates/topos-tce-storage/src/tests/checkpoints.rs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    assert_eq!(checkpoint.len(), 2);
+    assert_eq!(
+        *checkpoint.get(&SOURCE_SUBNET_ID_1).unwrap(),
+        (21..=31).collect::<Vec<_>>()
+    );
+    assert_eq!(
+        *checkpoint.get(&SOURCE_SUBNET_ID_2).unwrap(),
+        (0..=23).collect::<Vec<_>>()
+    );
+}
```
Comment (dvdplm) [inline crates/topos-tce-storage/src/tests/checkpoints.rs:123]: It would be good to have some more tests to make sure the code behaves the way we expect when things have gone wrong. Like, missing certificates, conflicting chains (some TCE node considers something delivered and some don't?) and other such unexpected scenarios. Can be different PR(s) too.

File: crates/topos-tce-storage/src/validator/mod.rs [hunk 6]
```diff
@@ -356,24 +359,24 @@ impl ValidatorStore {
...
                 if local_position <= position.delivery_position.position {
                     continue;
                 }
+
                 self.fullnode_store
                     .perpetual_tables
                     .streams
-                    .prefix_iter_at(&subnet, &position)?
+                    .prefix_iter(&(&subnet, &position.delivery_position.position))?
+                    .skip(1)
```
Comment (dvdplm) [inline crates/topos-tce-storage/src/validator/mod.rs:371]: Is this the "hack" we need to fix the bug? Have you figured out exactly why rocksdb behaves weirdly with `prefix_iter_at`? At Parity many years back we had issues with rocksdb iterators behaving in unexpected ways but I have forgotten the details tbh.

File: crates/topos-tce-synchronizer/src/lib.rs [hunk 7]
```diff
@@ -143,6 +144,12 @@ impl GrpcSynchronizerService for SynchronizerService {
         request: Request<CheckpointRequest>,
     ) -> Result<Response<CheckpointResponse>, Status> {
         let request = request.into_inner();
```
Comment (dvdplm) [inline crates/topos-tce-synchronizer/src/lib.rs:146]: I have a dumb question: what actually *is* a checkpoint? an you point me to the docs where it is described? The type is a vector of delivery proofs, which is what confuses me a bit I think.

File: crates/topos-tce-synchronizer/src/lib.rs [hunk 8]
```diff
@@ -151,41 +158,62 @@ impl GrpcSynchronizerService for SynchronizerService {
...
+                                    .into_iter()
+                                    .map(|(ready, signature)| SignedReady { ready, signature })
+                                    .collect(),
+                                threshold: v.threshold,
+                            })
+                            .collect();
+                        CheckpointMapFieldEntry {
+                            key: key.to_string(),
+                            value: v,
+                        }
```
Comment (dvdplm) [inline crates/topos-tce-synchronizer/src/lib.rs:196]: This looks like we're cloning the `ProofOfDelivery`s from the `diff` in order to create the `CheckpointMapFieldEntry`. Can't we just use `clone()`? I must be missing something. Is it the `source_subnet_id` and `delivery_position` the must be plucked out individually?

File: crates/topos-tce-synchronizer/src/checkpoints_collector/mod.rs [hunk 9]
```diff
@@ -184,11 +189,14 @@ impl CheckpointSynchronizer {
...
+        for (subnet, proofs) in diff {
             let len = proofs.len();
             let unverified_certs = self.store.insert_unverified_proofs(proofs)?;
 
-            debug!("Persist {} unverified proofs", len);
+            debug!(
+                "Persist {} unverified proof of delivery for {}",
+                len, subnet
+            );
             certs.extend(&unverified_certs[..]);
```
Comment (dvdplm) [inline crates/topos-tce-synchronizer/src/checkpoints_collector/mod.rs:200]:
```suggestion
            debug!("Persist {} unverified proofs for {}", proofs.len(), subnet);
```
Doesn't seemt to be used elesewhere?

File: crates/topos-tce-broadcast/src/double_echo/broadcast_state.rs [hunk 10]
```diff
@@ -230,13 +230,20 @@ impl BroadcastState {
...
             .subscriptions_view
             .network_size
             .checked_sub(self.subscriptions_view.ready.len())
         {
             Some(consumed) => consumed >= self.delivery_threshold,
             None => false,
-        }
+        };
+
+        debug!(
```
[the remark this answers is not in the document]
Comment (Freyskeyd) [reply crates/topos-tce-broadcast/src/double_echo/broadcast_state.rs:242]: If you consider this particular log as high traffic, I'm suggesting that we create a task to list more logs that are really more critical than this one.

File: crates/topos-tce-storage/src/tests/checkpoints.rs (hunk 4 above)
Comment (Freyskeyd) [reply crates/topos-tce-storage/src/tests/checkpoints.rs:84]: In this test, the `get_checkpoint_diff` method has an input of `&[checkpoint]` (line 100) while the `get_checkpoint_diff_with_no_input` `get_checkpoint_diff` method doesn't have one (line 57).

File: crates/topos-tce-storage/src/tests/checkpoints.rs (hunk 5 above)
Comment (Freyskeyd) [reply crates/topos-tce-storage/src/tests/checkpoints.rs:123]: Feel free to create a ticket to add them, as this PR is here to unlock the `testnet` I will not add more things to it.

File: crates/topos-tce-storage/src/validator/mod.rs [hunk 11]
```diff
@@ -336,14 +336,17 @@ impl ValidatorStore {
...
-            .map(|v| (v.delivery_position.subnet_id, vec![v]))
+        let from_positions: HashMap<SubnetId, &ProofOfDelivery> = from
+            .iter()
+            .map(|v| (v.delivery_position.subnet_id, v))
             .collect();
 
+        let mut output: HashMap<SubnetId, Vec<ProofOfDelivery>> =
+            from_positions.iter().map(|(s, _)| (*s, vec![])).collect();
+
         // Request the local head checkpoint
```
[the remark this answers is not in the document]
Comment (Freyskeyd) [reply crates/topos-tce-storage/src/validator/mod.rs:350]: It means the `head` of every source subnet `streams`, a `checkpoint` is a collection of `subnet's position`

File: crates/topos-tce-storage/src/validator/mod.rs (hunk 6 above)
Comment (Freyskeyd) [reply crates/topos-tce-storage/src/validator/mod.rs:371]: It is due to how we insert the prefix (`bincode` encoded fixed size int of subnet + position) and how we seek the key.

File: crates/topos-tce-synchronizer/src/checkpoints_collector/mod.rs (hunk 9 above)
Comment (Freyskeyd) [reply crates/topos-tce-synchronizer/src/checkpoints_collector/mod.rs:200]: I don't understand the suggestion

File: crates/topos-tce-synchronizer/src/lib.rs (hunk 7 above)
Comment (Freyskeyd) [reply crates/topos-tce-synchronizer/src/lib.rs:146]: The related RFC is [here](https://github.com/toposware/rfcs/blob/main/rfcs/0003-local-storage.md)

File: crates/topos-tce-synchronizer/src/lib.rs (hunk 8 above)
Comment (Freyskeyd) [reply crates/topos-tce-synchronizer/src/lib.rs:196]: This part is building the gRPC response based on the result of the query on the `store`.

File: crates/topos-tce-synchronizer/src/checkpoints_collector/mod.rs (hunk 9 above)
Comment (dvdplm) [reply crates/topos-tce-synchronizer/src/checkpoints_collector/mod.rs:200]: The suggestion is off, not sure what happened here, maybe some bug in the vscode PR review plugin. Anyway, the suggestion here was to remove the `let len = proofs.len()` line and just use `proofs.len()` in the debug statement. Just a nitpick really.
